Get argv from main entry point encoded as UTF-8

piponazo opened the pull request:
Fix #1996 

Thanks to the discussion we had ad #1996, I could find out how to properly get the arguments from argv on Windows properly encoded as UTF-8. Thanks to this, we can drop more than 1K LoC which was conditionally enabled by `EXV_UNICODE_PATH` (by setting the CMake option: EXIV2_ENABLE_WIN_UNICODE). 

I could not have found how to do this without checking this project:
https://github.com/huangqinjin/wmain

Other interesting article about the topic can be found here:
https://utf8everywhere.org/

piponazo committed:
Add system test to reproduce issue with special characters

piponazo committed:
windows: Trick to get UTF-8 encoded argv in main entry point

- Adapt exifprint to the new wmain strategy
- Delete have_unicode_path
- wmain does not work with MSYS & MinGW
- cmake: entry point via cmake instead of pragma
- cmake: better doc for MSVC flags
- Fix entry point in sample apps
- Adapt CMake code to work with MSVC & MinGW

piponazo committed:
Remove dead code around EXV_UNICODE_PATH

kmilos commented on cmake/compilerFlags.cmake:
```diff
@@ -169,6 +169,7 @@ if(MSVC)
 
     # Object Level Parallelism
     add_compile_options(/MP)
+    add_compile_options(/utf-8)
```
Do we need something equivalent for MinGW gcc/clang, or nothing is required there? I.e. I don't see `-DUNICODE -D_UNICODE` defined anywhere that would give us string-wstring etc. on MinGW, if presumably `/utf-8` does this for MSVC? If these are equivalent, maybe we prefer the compiler independent definitions?

Also maybe add a separate comment, the one above does not pertain to this line?

piponazo replied:
The MSYS2/MinGW CI jobs indeed failed. I have to check locally with that environment to check what to do there. The new test added is at the moment failing on that platform:
https://github.com/Exiv2/exiv2/runs/5154900058?check_suite_focus=true

kmilos replied:
I personally would like to have a portable codebase that runs on Windows without assuming the compiler. There are many projects out there that rely on MinGW provided binaries.

piponazo replied:
Me too. I would like to take a pragmatic approach and fix one thing at each time. We can focus on this PR on fixing #1996 for MSVC and later investigate how to do it with MinGW.

I think that by following that approach the build with MinGW would remain the same (we would not be breaking anything).

However, I need to check how I could skip the new test on Win+MinGW builds

kmilos replied:
I don't think this is the way to go then if this is MSVC specific.

I still don't get why this approach won't (or currently doesn't) work:

- have main() w/ char* argc (i.e. assume multi-byte UTF-8 input for all arguments on all platforms)
- parse arguments as multi-byte UTF-8 char* only on all platforms (i.e. as today)
- on Windows, convert **only filenames** w/ existing s2ws() when needed to open the I/O w/ Windows wchar APIs

The only sticking point, as you say, might be what the different consoles encode, I haven't checked that thoroughly.

kmilos reviewed (commented):
Very nice!

piponazo commented on src/exiv2.cpp:
```diff
@@ -120,6 +120,8 @@ namespace {
 // Main
 int main(int argc, char* const argv[])
 {
+    setlocale(LC_CTYPE, ".utf8");
```
Good point, yes, indeed Windows 10 is the minimum. We either say: Only Win10 or greater is supported in the command line application or we add some preprocessor conditions there.

clanmills replied:
It's reasonable for the exiv2 command-line program to be targeted at Windows/10 and later.

Exiv2 "legacy support" is branch `0.27-maintenance` and  branch 'main' is about modernisation.  If a user needs something "modern" to run on "legacy", they can report an issue and a PR.

clanmills commented on src/wmain.c:
```diff
@@ -0,0 +1,43 @@
+#include <windows.h>
+
+// https://docs.microsoft.com/en-us/cpp/build/reference/entry-entry-point-symbol
+#pragma comment(linker, "/ENTRY:wWinMainCRTStartup")
```
I would discourage you from using `pragma comment(linker....`.  It's "hidden magic".  I would prefer to see this explicitly declared in the CMake code.

piponazo commented on src/version.cpp:
```diff
@@ -437,10 +437,6 @@ void Exiv2::dumpLibraryInfo(std::ostream& os,const std::vector<std::regex>& keys
 #endif
 
-#ifdef EXV_UNICODE_PATH
-     have_unicode_path=1;
```
Yes, they were removed at 821d6c3fb9063741c038c6d260755fc7be52a34e

clanmills commented:
I found running exiv2/msvc builds work fine on DOS.  exiv2/gcc/mingw builds work fine on MinGW/bash.  And that's what we build and test.

Trying to use msvc/utf-8 enabled applications on MinGW/bash is a horror show which has nothing to do with exiv2.

kmilos commented:
I also wonder if our current s2ws()/ws2s() implementation is entirely correct and perhaps shadowing some problems here - most conversions like this use CP_UTF8 and not CP_ACP here, including [MinGW's own helper function](https://github.com/Alexpux/mingw-w64/blob/d0d7f784833bbb0b2d279310ddc6afb52fe47a46/mingw-w64-crt/misc/mingw_mbwc_convert.c#L26-L43).

piponazo commented:
> > Trying to use msvc/utf-8 enabled applications on MinGW/bash is a horror show which has nothing to do with exiv2.
> 
> However, the inverse should/must be possible. MinGW built programs are as native Windows programs as any, and should run in cmd and PowerShell just like any other Windows program.

Right, I normally can run MSVC builds from any terminal (cmd, powershell, bash). With the MinGW build it should be the same. The only difference at the moment is that with MSVC we can use the flag `/utf-8` to indicate that the execution character set is UTF-8:
https://docs.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170

At the moment I did not find a way to make this work with MinGW.

neheb commented:
I doubt this can ever workeith MinGW. MSYS has a UCRT and clang tool chain that should work fine.

kmilos commented:
> I doubt this can ever workeith MinGW. MSYS has a UCRT and clang tool chain that should work fine.

Eh? When we say MinGW we mean [MSYS2 MINGW64/UCRT64/CLANG64 (and hopefully CLANGARM64 soon)](https://www.msys2.org/docs/environments/). Plenty of UTF-8 programs there for years w/ MSVCRT. This is like saying there was no UTF-8 before UCRT, like no MSVC compiled programs against MSVCRT before Windows 8 and VS2017 were UTF-8 capable?!

neheb commented:
> > I doubt this can ever workeith MinGW. MSYS has a UCRT and clang tool chain that should work fine.
> 
> Eh? When we say MinGW we mean [MSYS2 MINGW64/UCRT64/CLANG64 (and hopefully CLANGARM64 soon)](https://www.msys2.org/docs/environments/). Plenty of UTF-8 programs there for years w/ MSVCRT. This is like saying there was no UTF-8 before UCRT, like no MSVC compiled programs against MSVCRT before Windows 8 and VS2017 were UTF-8 capable?!

I don't think I've seen utf-8 programs with msvcrt. I've seen plenty of utf-16 ones but not utf-8.

clanmills commented:
I'm not sure we need to do anything here. When this topic was brought up a few months ago, we discovered that the posix platforms (mingw, macOS, Linux) required no attention.  The aim is to get the Greek Path to work on msvc/Windows and that's what you now have working.  Great Job.

@kmilos.  Milo&scaron; your point about `s2ws()/ws2s()` is interesting, however I don't believe that's related to the Greek Path.

piponazo commented:
@clanmills The issue #1996 correctly states that he problem is present on MSVC & MinGW builds.

With this PR I am proposing to fix the MSVC build and discard all the usages of `std::wstring` and `wchar_t`. By doing so, we will start to assume that `std::string` and `char*` variables are considered UTF-8, anywhere in the program. That seems to be enough thanks to the new UTF-8 support on the Universal C Runtime (UCRT) which Windows 10 brings.

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170#utf-8-support

The problem seems to be that MinGW does not (and cannot) use UCRT:
https://mingwpy.github.io/ucrt.html
https://stackoverflow.com/questions/57528555/how-do-i-build-against-the-ucrt-with-mingw-w64

I think we could merge this as it is, solve 1 of the 2 problems (MSVC build) and if somebody else is interested in the topic, we could work on also fixing the MinGW problem in another branch.

neheb commented:
Note that UCRT is available for 7 and 8.1. So it's not that bad.

kmilos commented:
> The problem seems to be that MinGW does not (and cannot) use UCRT

[Note the link above again: MSYS2 has many environments](https://www.msys2.org/docs/environments/) for which exiv2 is built. At least UCRT64 and CLANG64 link to UCRT.

kmilos commented:
> Note that UCRT is available for 7 and 8.1. So it's not that bad.

You might be right about MSVCRT, sorry. From [here](https://developer.r-project.org/Blog/public/2020/05/02/utf-8-support-on-windows/index.html): 

>UCRT can be installed on older versions of Windows, but UTF-8 support will only work on Windows 10 (November 2019 update) and newer.

piponazo commented:
I found some interesting thread in which MinGW people is discussing about provided UCRT version of their packages:
https://github.com/msys2/MINGW-packages/issues/6901

So it seems that at some point we will be able to compile with MSYS2/MinGW and link against UCRT 🛩️

kmilos commented:
> So it seems that at some point we will be able to compile with MSYS2/MinGW and link against UCRT

That point **was** as soon as UCRT64 environment was made available 😉 

Exiv2 builds are available for all supported MSYS2 environments: https://packages.msys2.org/base/mingw-w64-exiv2

piponazo commented:
After I followed the link provided by @kmilos I tried to use the packages which are using UCRT:
https://www.msys2.org/docs/environments/

Exiv2 compiled correctly, I got some some pop-ups complaining about missing DLLs, I placed them in the bin directory and I finally ended up with a weird crash 😭 

![image](https://user-images.githubusercontent.com/102645/153637580-82b2e1bb-f8e4-460a-b535-ba63ae7cbf40.png)

neheb commented:
strange.

piponazo committed:
Replace platform specific code with std::filesystem

piponazo committed:
ci: Use MSYS UCRT environment

- Use the CMake generator 'MSYS Makefiles' for MSYS builds
- Run CI build in parallel
- MSYS with NLS OFF

postscript-dev commented:
This is a great improvement for non-English speakers.


Is it also worth checking that the lens configuration file works with a Greek path?

piponazo commented:
I have good and bad news.

Let's start with the good ones. I could finally find out how to compile with MSYS_MinGW (UCRT) properly. What I was missing is to add the `-municode` flag for both: compilation and linking. More details in this commit: b37842cb5a2261ff408367d73a25230725c501fe .

When doing that, some of the functions making use of the UCRT would complain about obtaining `char*` instead of `wchar*`, for example: `GetFullPathName` or `GetTempPath`. However, since we can now use C++17, I replaced those conflictive parts of the code by the usage of `std::filesystem`: 1fe8066d129b5b2909025da512a1264d41316e99

Actually thanks to the previous commit, I have discovered some other areas of the code that could be simplified by removing platform specific code and replace it by `std::filesystem`. 

The bad news: I adapted the CI job to use MSYS_MinGW with the UCRT environment, and it builds correctly. However, it looks like at the stage of running the tests, the jobs are getting stuck:
https://github.com/Exiv2/exiv2/actions/runs/1836200447

Interestingly, when testing things locally, I need to be in a MSYS(UCRT) terminal so that the generated **exiv2** application can find properly the DLLs available in the UCRT environment. If I try to run that binary from a CMD or PowerShell, I get either pop-ups complaining about missing DLLs, or the **exiv2** app simply do not run. Probably the CI jobs are suffering of this effect ... 

I'll keep trying to fix the MSYS(UCRT) CI jobs.

piponazo committed:
doc: update notes about MSYS

piponazo commented:
> Is it also worth checking that the lens configuration file works with a Greek path?

I know very little about the lens part of the code. I would propose that if somebody wants to check that, please do it as a separate task in another branch. I already dedicated many hours to this topic and the PR if of considerable size.

piponazo committed:
Remove unused PRINTUCS2 CMake option

postscript-dev commented:
> I know very little about the lens part of the code. I would propose that if somebody wants to check that, please do it as a separate task in another branch. I already dedicated many hours to this topic and the PR if of considerable size.


@piponazo : No problem, this can be looked at in a different PR.

This branch looks like an inteligent solution to the Unicode problem.

kmilos commented:
Good going Luis. Sorry I wasn't able to help because I have all these environments set up on my box already, but I am unfortunately away for a week...

piponazo committed:
Apply suggestions from code review

Co-authored-by: Kevin Backhouse <kevinbackhouse@github.com>

piponazo commented on src/wmain.c:
```diff
@@ -0,0 +1,40 @@
+int wmain(int argc, wchar_t* argv[])
+{
+    char** args;
+    int nbytes = sizeof(char*) * (argc + 1);
```
I had to revert this one, windows did not like to make an implicit cast later from `size_t` to `int`. Let's leave it as I found this file from this repository: https://github.com/huangqinjin/wmain

hassec commented on .github/workflows/on_PR_windows_matrix.yml:
```diff
@@ -91,59 +90,59 @@ jobs:
         if: ${{matrix.platform == 'x64'}}
         run: |
-          ctest --test-dir build --output-on-failure
+          ctest --test-dir build --output-on-failure --stop-on-failure
```
Would this stop the testing when it encounters the first failing test?       
I think I prefer if the ci shows me all failing tests, not just the first one.

piponazo replied:
Yes, I knew this could generate some controversy 😇 . I added it because at some point the tests on MSYS were taking more than 1 hour and I wanted to break as soon as possible.

It is fine for me to revert these changes now that everything is working.

piponazo committed:
ci: do not stop on failure

clanmills commented:
Great Work, @piponazo.  Well Done.
